proxy/healthy_endpoints_test.go

```diff
@@ -150,142 +161,84 @@ func TestPHCForSingleHealthyEndpoint(t *testing.T) {
-			count200, ok := va.CountStatus(200)
-			assert.True(t, ok)
-			assert.InDelta(t, count200, 15000, 50) // 3000*5s, the delta is for CI
-			assert.Equal(t, count200, int(va.TotalRequests()))
-		})
-	}
-
-	t.Run("consistent hash with balance factor", func(t *testing.T) {
-		_, ps := setupProxy(t, fmt.Sprintf(`* -> consistentHashKey("${request.header.ConsistentHashKey}") -> consistentHashBalanceFactor(1.25) -> <consistentHash, "%s", "%s", "%s">`,
-			services[0], services[1], services[2]))
-		va := fireVegeta(t, ps, 3000, 1*time.Second, 5*time.Second)
-		count200, ok := va.CountStatus(200)
-		assert.True(t, ok)
-		assert.Equal(t, count200, int(va.TotalRequests()))
-	})
-}
-
-func TestPHCForMultipleHealthyAndOneUnhealthyEndpoints(t *testing.T) {
-	services := setupServices(t, 2, 1)
-	for _, algorithm := range []string{"random", "roundRobin", "powerOfRandomNChoices"} {
-		t.Run(algorithm, func(t *testing.T) {
-			mockMetrics, ps := setupProxy(t, fmt.Sprintf(`* -> backendTimeout("5ms") -> consistentHashKey("${request.header.ConsistentHashKey}") -> <%s, "%s", "%s", "%s">`,
-				algorithm, services[0], services[1], services[2]))
-
-			va := fireVegeta(t, ps, 3000, 1*time.Second, 5*time.Second)
-
-			count200, ok := va.CountStatus(200)
-			assert.True(t, ok)
-
-			count504, ok := va.CountStatus(504)
-			assert.True(t, ok)
-
-			failedRequests := math.Abs(float64(va.TotalRequests())) - float64(count200)
-			t.Logf("total requests: %d, count200: %d, count504: %d, failedRequests: %f", va.TotalRequests(), count200, count504, failedRequests)
-
-			assert.InDelta(t, float64(count504), failedRequests, 5)
-			assert.InDelta(t, 0, float64(failedRequests), 0.1*float64(va.TotalRequests()))
-			mockMetrics.WithCounters(func(counters map[string]int64) {
-				assert.InDelta(t, float64(va.TotalRequests()), float64(counters["passive-health-check.endpoints.dropped"]), 0.3*float64(va.TotalRequests())) // allow 30% error
+func TestPHC(t *testing.T) {
+	for _, tt := range []struct {
+		name               string
+		healthy, unhealthy int
+	}{
+		{"single healthy", 1, 0},
+		{"multiple healthy", 2, 0},
+		{"multiple healthy and one unhealthy", 2, 1},
+		{"multiple healthy and multiple unhealthy", 2, 2},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			services := setupServices(t, tt.healthy, tt.unhealthy)
+			servicesString := formatServicesString(services)
+			for _, algorithm := range []string{"random", "roundRobin", "powerOfRandomNChoices"} {
+				t.Run(algorithm, func(t *testing.T) {
+					mockMetrics, ps := setupProxy(t, fmt.Sprintf(`* -> backendTimeout("5ms") -> consistentHashKey("${request.header.ConsistentHashKey}") -> <%s, %s>`,
+						algorithm, servicesString))
+
+					va := fireVegeta(t, ps, 3000, 1*time.Second, 5*time.Second)
+
+					count200, ok := va.CountStatus(200)
+					assert.True(t, ok)
+
+					count504, ok := va.CountStatus(504)
+					assert.Condition(t, func() bool {
+						if tt.unhealthy == 0 && (ok || count504 == 0) {
```

Comment:
Could you kindly describe in English language (here or in comment to this code), what this function should check?

Reply:
if we don't have unhealthy endpoints, then we don't have 504s this mean that `assert.True` would fail which we don't want. The condition here check if we have don't have unhealthy endpoints then the metric shouldn't find the key for 504 or at least return 0, the other case can be simplified to `return count504 > 0`, does this make more sense?

Reply:
Ah, I see, in other words, this condition checks:

+ if there're no unhealthy endpoints: check whether `ok` or `count504`
+ if there's at least one unhealthy endpoint: check `ok` only

Okay, now makes sense to me.